packages/itwinui-react/src/utils/hooks/useControlledState.ts

```diff
@@ -27,14 +30,40 @@ export const useControlledState = <T>(
     () => (controlledState !== undefined ? controlledState : uncontrolledState),
     [controlledState, uncontrolledState],
   );
+  const oldState = React.useRef<T>(state);
 
   const setState = React.useCallback(
     (value) => {
+      const newValue = (() => {
+        if (typeof value === 'function') {
+          return (value as (prevState: T) => T)(state);
+        }
+        return value;
+      })();
+
+      // Called only when the value *changes* (uncontrolled mode) or should *change* (controlled mode).
+      if (newValue === oldState.current) {
+        return;
+      }
+
+      // If in controlled mode, defer updating oldState to the sync effect.
+      // Else, update oldState here since uncontrolledState is guaranteed to be set to value.
+      if (controlledState == null) {
+        oldState.current = newValue;
```

Comment:
Not sure I understand why this needs to be split across here and the Effect. The comments do not help.

Reply:
A code comment concisely describing _why_ this line is needed would be helpful.

e.g.

```js
// Immediately update oldValue, in case setState is called
// multiple times during the same render.
oldState.current = newValue;
```

(I'm not sure if this explanation is accurate though)